internal/app/machined/pkg/controllers/block/volume_config.go

```diff
@@ -0,0 +1,287 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package block
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"github.com/siderolabs/go-blockdevice/v2/encryption"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/internal/pkg/meta"
+	"github.com/siderolabs/talos/internal/pkg/partition"
+	"github.com/siderolabs/talos/pkg/machinery/cel"
+	"github.com/siderolabs/talos/pkg/machinery/cel/celenv"
+	cfg "github.com/siderolabs/talos/pkg/machinery/config/config"
+	"github.com/siderolabs/talos/pkg/machinery/config/types/v1alpha1"
+	"github.com/siderolabs/talos/pkg/machinery/constants"
+	"github.com/siderolabs/talos/pkg/machinery/resources/block"
+	"github.com/siderolabs/talos/pkg/machinery/resources/config"
+	"github.com/siderolabs/talos/pkg/machinery/resources/runtime"
+)
+
+// VolumeConfigController provides volume configuration based on Talos defaults and machine configuration.
+type VolumeConfigController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *VolumeConfigController) Name() string {
+	return "block.VolumeConfigController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *VolumeConfigController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
+			ID:        optional.Some(config.V1Alpha1ID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: runtime.NamespaceName,
+			Type:      runtime.MetaKeyType,
+			ID:        optional.Some(runtime.MetaKeyTagToID(meta.StateEncryptionConfig)),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *VolumeConfigController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: block.VolumeConfigType,
+			Kind: controller.OutputShared,
+		},
+	}
+}
+
+func labelVolumeMatch(label string) cel.Expression {
+	return cel.MustExpression(cel.ParseBooleanExpression(fmt.Sprintf("volume.partition_label == '%s'", label), celenv.VolumeLocator()))
+}
+
+func labelVolumeMatchAndNonEmpty(label string) cel.Expression {
+	return cel.MustExpression(cel.ParseBooleanExpression(fmt.Sprintf("volume.partition_label == '%s' && volume.name != ''", label), celenv.VolumeLocator()))
+}
+
+func metaMatch() cel.Expression {
+	return cel.MustExpression(cel.ParseBooleanExpression(fmt.Sprintf("volume.partition_label == '%s' && volume.name in ['', 'talosmeta']", constants.MetaPartitionLabel), celenv.VolumeLocator()))
```

Comment:
why matching on `''` string?

Reply:
the `META` is initially created on empty, so before the first write it will be all zeroes, and its type will be `""` (nothing)